src/parser.ts

```diff
@@ -187,9 +183,197 @@ function getFunctionName(func: ts.FunctionLikeDeclaration): string {
+    let member;
+    if (
+        ts.isMethodDeclaration(func) ||
+        ts.isGetAccessor(func) ||
+        ts.isSetAccessor(func) ||
+        ts.isConstructorDeclaration(func)
+    ) {
+        member = func;
+    } else if (
+        ts.isFunctionExpression(func) ||
+        ts.isArrowFunction(func) ||
+        ts.isClassExpression(func)
+    ) {
+        if (
+            ts.isPropertyDeclaration(func.parent) ||
+            ts.isPropertyAssignment(func.parent)
+        ) {
+            member = func.parent;
+            isProperty = true;
+        }
+    }
+    return { member, isProperty };
+}
+
+function getObjectLiteralPrefix(
+    func: ts.FunctionLikeDeclaration | ts.ClassLikeDeclaration
+): string | null {
+    let { member, isProperty } = getMemberNodeForFunction(func);
+    // member === null means the function was neither a method nor a
+    // property, so it is not part of an object literal
+    if (!member) {
+        return null;
+    }
+    // if the function is a property, and either has a local name,
+    // or is inside a class expression, we take the local name and
+    // and do not prefix it with the object literal chain
+    if (isProperty) {
+        if (func.name !== undefined || ts.isClassExpression(member.parent)) {
+            return null;
+        }
+    }
+    return getObjectLiteralPrefixR(member.parent);
+}
+
+function getObjectLiteralPrefixR(
+    node: ts.Node,
+    prefixWithObject = false
+): string | null {
+    const parent = node.parent;
+    if (!parent) {
+        return null;
+    }
+    if (ts.isPropertyAssignment(parent)) {
+        const propertyAssignment = parent;
+        const prefix = getObjectLiteralPrefixR(
+            propertyAssignment.parent,
+            true
+        );
+        return `${prefix}.${getPropertyName(propertyAssignment.name)}`;
+    } else if (ts.isVariableDeclaration(parent)) {
+        if (ts.isIdentifier(parent.name)) {
+            return getPropertyName(parent.name);
+        }
+    } else if (
+        ts.isBinaryExpression(parent) &&
+        parent.operatorToken.kind === ts.SyntaxKind.EqualsToken
+    ) {
+        if (ts.isIdentifier(parent.left)) {
+            return getPropertyName(parent.left);
+        } else if (
+            ts.isPropertyAccessExpression(parent.left) ||
+            ts.isElementAccessExpression(parent.left)
+        ) {
+            return getLeftHandSideName(parent.left);
+        }
+    }
+    return prefixWithObject ? "<Object>" : null;
+}
+
+function getParentClassProperty(func: ts.FunctionLikeDeclaration) {
+    let { member } = getMemberNodeForFunction(func);
+    if (member) {
+        const parent = member.parent;
+        if (ts.isClassDeclaration(parent) || ts.isClassExpression(parent)) {
+            return parent;
+        }
+    }
+    return null;
+}
+
+function getClassPrefix(func: ts.FunctionLikeDeclaration): string | null {
+    let { member, isProperty } = getMemberNodeForFunction(func);
+    if (member) {
```

Comment:
Might be slightly easier to read if we do early exit by inverting the condition. This reduces indentation.